python/src/iceberg/schema.py

```diff
@@ -0,0 +1,448 @@
+
+    @property
+    def identifier_field_ids(self) -> List[int]:
+        return self._identifier_field_ids
+
+    def _lazy_id_to_field(self) -> Dict[int, NestedField]:
+        """Returns an index of field ID to NestedField instance
+
+        This property is calculated once when called for the first time. Subsequent calls to this property will use a cached index.
+        """
+        if not self._id_to_field:
+            self._id_to_field = index_by_id(self)
+        return self._id_to_field
+
+    def _lazy_name_to_id_lower(self) -> Dict[str, int]:
+        """Returns an index of lower-case field names to field IDs
+
+        This property is calculated once when called for the first time. Subsequent calls to this property will use a cached index.
+        """
+        if not self._name_to_id_lower:
+            self._name_to_id_lower = {name.lower(): field_id for name, field_id in self._name_to_id.items()}
+        return self._name_to_id_lower
+
+    def as_struct(self) -> StructType:
+        """Returns the underlying struct"""
+        return self._struct
+
+    def find_field(self, name_or_id: str | int, case_sensitive: bool = True) -> NestedField:
+        """Find a field using a field name or field ID
+
+        Args:
+            name_or_id (str | int): Either a field name or a field ID
+            case_sensitive (bool, optional): Whether to peform a case-sensitive lookup using a field name. Defaults to True.
+
+        Returns:
+            NestedField: The matched NestedField
+        """
+        if isinstance(name_or_id, int):
+            field = self._lazy_id_to_field().get(name_or_id)
+            return field  # type: ignore
+        if case_sensitive:
+            field_id = self._name_to_id.get(name_or_id)
+        else:
+            field_id = self._lazy_name_to_id_lower().get(name_or_id.lower())
+        return self._lazy_id_to_field().get(field_id)  # type: ignore
+
+    def find_type(self, name_or_id: str | int, case_sensitive: bool = True) -> IcebergType:
+        """Find a field type using a field name or field ID
+
+        Args:
+            name_or_id (str | int): Either a field name or a field ID
+            case_sensitive (bool, optional): Whether to peform a case-sensitive lookup using a field name. Defaults to True.
+
+        Returns:
+            NestedField: The type of the matched NestedField
+        """
+        field = self.find_field(name_or_id=name_or_id, case_sensitive=case_sensitive)
+        return field.type  # type: ignore
+
+    def find_column_name(self, column_id: int) -> str:
+        """Find a column name given a column ID
+
+        Args:
+            column_id (int): The ID of the column
+
+        Raises:
+            ValueError: If no column name can be found for the given column ID
+
+        Returns:
+            str: The column name
+        """
+        column = self._lazy_id_to_field().get(column_id)
+        return None if column is None else column.name  # type: ignore
```

Comment:
This actually needs to return the full name, not the field name. In Java, this uses the same index visitor as the name to ID, but it produces the `byId` map.